File: Phase1/src/main/java/InvertedIndex.java
import java.util.HashMap;
import java.util.Map;
import java.util.TreeSet;

public class InvertedIndex {
    HashMap<String, String> allDocuments = new HashMap<>();
    HashMap<String, TreeSet<String>> tokenizedWords = new HashMap<>();

    public void tokenizeFiles (String folderName){
        FileReader fileReader = new FileReader();
        allDocuments = fileReader.readingFiles(folderName);


        for (Map.Entry<String, String> entry : allDocuments.entrySet()) {
            String value = entry.getValue();
            TreeSet<String> rawAllWords = tokenize(value);
            TreeSet<String> allWords = StringUtils.processRawTokens(rawAllWords);
            for (String allWord : allWords) {
                if (tokenizedWords.containsKey(allWord)) {
                    tokenizedWords.get(allWord).add(entry.getKey());
                } else {
                    TreeSet<String> docks = new TreeSet<>();
                    docks.add(entry.getKey());
                    tokenizedWords.put(allWord.toLowerCase(), docks);
                }
            }
        }
    }

    private TreeSet<String> tokenize(String string) {
        TreeSet<String> tokens = new TreeSet<>();
        for (int i = 0; i < string.length(); i++) {
            if (!Character.isAlphabetic(string.charAt(i)))
                continue;
            StringBuilder token = new StringBuilder();
            while (i < string.length() && Character.isAlphabetic(string.charAt(i))) {
                token.append(Character.toLowerCase(string.charAt(i)));
                i++;
            }
            tokens.add(token.toString());
        }
        return tokens;
    }

    public TreeSet<String> query(String input) {
        UserInput userInput = new UserInput(input);
        TreeSet<String> result = null;
        for (String string : userInput.getAndInputs()) {
            result = andWordWithResult(string, result);
        }
        for (String string : userInput.getOrInputs()) {
            result = addWordToResult(string, result);
        }
        for (String string : userInput.getRemoveInputs()) {
            result = removeWordFromResult(string, result);
        }
        if (result != null)
            return result;
        else
            return new TreeSet<>();
    }

    private TreeSet<String> removeWordFromResult(String word, TreeSet<String> result) {
        TreeSet<String> wordList = getDocsContainWord(word);
        if (result == null) {
            return new TreeSet<>();
        }
        result.removeAll(wordList);
        return result;
    }

    private TreeSet<String> addWordToResult(String word, TreeSet<String> result) {
        TreeSet<String> wordList = getDocsContainWord(word);
        if (result == null) {
            return (TreeSet<String>) wordList.clone();
        }
        result.addAll(wordList);
        return result;
    }

    private TreeSet<String> andWordWithResult(String word, TreeSet<String> result) {
        TreeSet<String> wordList = getDocsContainWord(word);
        if (result == null) {
            return (TreeSet<String>) wordList.clone();
        }
        result.retainAll(wordList);
        return result;
    }

    private TreeSet<String> getDocsContainWord(String word) {
        if (tokenizedWords.containsKey(word))
            return tokenizedWords.get(word);
        else
            return new TreeSet<>();
    }
}

File: Phase5/SearchEngine/SearchEngine/InvertedIndex.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SearchEngine.Interfaces;

namespace SearchEngine
{
    public class InvertedIndex : IInvertedIndex
    {
        public Dictionary<string, SortedSet<string>> TokenizedWords { get; } = new();


        public IInvertedIndex TokenizeFiles(Dictionary<string, string> allDocuments)
        {
            allDocuments.ToList().ForEach(pair =>
            {
                foreach (var word in StringUtils.ProcessRawTokens(Tokenize(pair.Value)).ToList())
                {
                    if (TokenizedWords.ContainsKey(word))
                    {
                        TokenizedWords[word].Add(pair.Key);
                    }
                    else
                    {
                        SortedSet<string> docks = new SortedSet<string> { pair.Key };
                        TokenizedWords.Add(word.ToLower(), docks);
                    }
                }
            });

            return this;
        }

        private static SortedSet<string> Tokenize(string st)
        {
            SortedSet<string> tokens = new SortedSet<string>();
            for (int i = 0; i < st.Length; ++i)
            {
                if (!char.IsLetter(st[i])) continue;
                StringBuilder token = new StringBuilder();
                while (i < st.Length && char.IsLetterOrDigit(st[i]))
                {
                    token.Append(char.ToLower(st[i]));
                    i++;
                }

                tokens.Add(token.ToString());
            }

            return tokens;
        }

        public SortedSet<string> Query(IUserInput input)
        {
            SortedSet<string> result = null;
            input.GetAndInputs().ToList().ForEach(st => result = AndWordWithResult(st, result));
            input.GetOrInputs().ToList().ForEach(st => result = AddWordToResult(st, result));
            input.GetRemoveInputs().ToList().ForEach(st => result = RemoveWordFromResult(st, result));
            if (result != null) return result;
            else return new SortedSet<string>();
        }

        private SortedSet<string> RemoveWordFromResult(string word, SortedSet<string> result)
        {
            SortedSet<string> wordList = GetDocsContainWord(word);
            if (result == null) return new SortedSet<string>();
            result.ExceptWith(wordList);
            return result;
        }

        private SortedSet<string> AddWordToResult(string word, SortedSet<string> result)
        {
            SortedSet<string> wordList = GetDocsContainWord(word);
            if (result == null) return CloneSortedSet(wordList);
            result.UnionWith(wordList);
            return result;
        }

        private SortedSet<string> AndWordWithResult(string word, SortedSet<string> result)
        {
            SortedSet<string> wordList = GetDocsContainWord(word);
            if (result == null) return CloneSortedSet(wordList);
            result.IntersectWith(wordList);
            return result;
        }

        private SortedSet<string> GetDocsContainWord(string word)
        {
            if (TokenizedWords.ContainsKey(word)) return TokenizedWords[word];
            return new SortedSet<string>();
        }

        private SortedSet<string> CloneSortedSet(SortedSet<string> toClone)
        {
            SortedSet<string> cloneable = new SortedSet<string>();
            toClone.ToList().ForEach(st => cloneable.Add(st));
            return cloneable;
        }
    }
}